Add Media Uploader component

joshuatf opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adds a component for uploading media via button as well as drag and drop.

Closes # .

### How to test the changes in this Pull Request:

1. Run `npm run storybook`
2. Check out the various examples and examine media functionality (note that the WP media modal does not work here since we are outside the WP context)
3. Add the component to any WC React page with media scripts enqueued (should be most WCA pages).
```
<MediaUploader
    onSelect={ ( file ) => console.log( file ) }
    onError={ ( error ) =>console.log( error ) }
    onUpload={ ( files ) => console.log( files ) }
/>
```
4. Check that the WP media modal opens correctly and allows you to choose an image.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Commit:
Add media uploader component

Commit:
Expand on mocked examples in storybook

Commit:
Add styling

joshuatf committed:
Move MediaUploder out of index

joshuatf committed:
Add changelog entry

joshuatf commented on packages/js/components/src/media-uploader/media-uploader.tsx:
```diff
@@ -0,0 +1,78 @@
+				allowedTypes={ allowedMediaTypes }
+				render={ ( { open } ) => (
+					<Button variant="secondary" onClick={ open }>
+						{ buttonText }
```
Not sure how I feel about this yet, but we could optionally allow passing of a `render` prop instead of `buttonText` to thin out props a bit.

jacob-sewell replied:
I'd rather **have** two simple props than a function prop any day, but I hear what you're saying.

_Edited for word choice._

joshuatf commented on packages/js/components/src/media-uploader/media-uploader.tsx:
```diff
@@ -0,0 +1,78 @@
+	hasDropZone = true,
+	label = __( 'Drag images here or click to upload', 'woocommerce' ),
+	maxUploadFileSize = 10000000,
+	MediaUploadComponent = MediaUpload,
```
Note this component and `uploadMedia` both have defaults which typically work, but are allowed as props for dynamic injection and testing in StoryBook.

This may also prove useful if a user wants to replace the instance of the `MediaUpload` component.

jacob-sewell commented on packages/js/components/src/media-uploader/stories/index.tsx:
```diff
@@ -0,0 +1,164 @@
+		<Card size="large">
+			<CardBody>
+				{ error && (
+					<Notice status={ 'error' }>{ error.message }</Notice>
```
This story's error message doesn't dismiss on clicking the X icon. I don't know if this is just a problem in the Storybook context or if Notice doesn't provide a default that works....

joshuatf replied:
Removed the dismissable option for the notice.

jacob-sewell commented on packages/js/components/src/media-uploader/stories/index.tsx:
```diff
@@ -0,0 +1,164 @@
+					onSelect={ ( file ) => setImages( [ ...images, file ] ) }
+					onError={ () => null }
+					onUpload={ ( files ) =>
+						setImages( [ ...images, ...files ] )
```
When I select multiple files and drag them in, I only get one file in the `files` array. This does appear to just be an issue with this story, because when I test in a WooCommerce admin page I get multiple firings of this callback, which eventually includes all three files I dragged.

joshuatf replied:
Ya, I was being lazy by not wanting to mock out a full uploader, but went ahead and fleshed it out for clarity.

jacob-sewell commented on packages/js/components/src/media-uploader/media-uploader.tsx:
```diff
@@ -0,0 +1,78 @@
+				onSelect={ onSelect }
+				allowedTypes={ allowedMediaTypes }
+				render={ ( { open } ) => (
+					<Button variant="secondary" onClick={ open }>
```
I'm getting a TypeScript error on this line because apparently it doesn't think Button takes a variant prop. This is currently keeping my turbo build from completing.

jacob-sewell replied:
Having removed the variant prop, I was able to test in the WooCommerce admin home screen.

joshuatf replied:
Updated 👍

jacob-sewell reviewed (commented):
I found some issues with the Storybook stories, but in the real admin context it tests well, except for the TypeScript error that keeps me from building it.

joshuatf committed:
Handle PR feedback

joshuatf committed:
Add readme

joshuatf committed:
Fix missing media utils types

joshuatf commented:
Thanks for the feedback, @jacob-sewell!

I think I've addressed everything and also added a readme.  Could you take another look when you have a moment?

jacob-sewell commented on packages/js/components/src/media-uploader/media-uploader.tsx:
```diff
@@ -0,0 +1,78 @@
+				onSelect={ onSelect }
+				allowedTypes={ allowedMediaTypes }
+				render={ ( { open } ) => (
+					<Button isSecondary onClick={ open }>
```
On trunk this package now includes the current version of `@types/wordpress__components`, so we can stop using the deprecated `isSecondary` prop in favor of the currently-recommended `variant`.

Probably instead of holding this PR up to change that, we should have a low-priority issue to look for places where we're using the old Button props and change them all at once.

joshuatf replied:
Went ahead and updated this to make tests happy.

jacob-sewell reviewed (dismissed):
Let's get that moon! 🚀

joshuatf committed:
Rebase and use variant instead of isSecondary for Button

joshuatf committed:
Fix up lock file

jacob-sewell reviewed (dismissed):
Looks good!

joshuatf committed:
Fix up lint errors

jacob-sewell commented on packages/js/components/src/media-uploader/media-uploader.tsx:
```diff
@@ -12,7 +12,7 @@ import {
 } from '@wordpress/media-utils';
 
-const DEFAULT_ALLOWED_MEDIA_TYPES = ['image'];
+const DEFAULT_ALLOWED_MEDIA_TYPES = [ 'image' ];
```
Oh thank goodness, this whitespace problem was going to keep me up tonight.

jacob-sewell reviewed (approved):
I'm surprised my local environment didn't yell at me about the linting errors, but I'm glad to see them resolved.
